[CanonicalizeOSSALifetime] Run on lexical lifetimes.

nate-chandler opened the pull request:
Previously, the utility bailed out on lexical lifetimes because it didn't respect deinit barriers.  Here, deinit barriers are found and added to liveness if the value is lexical.  This enables copies to be propagated without hoisting destroys over deinit barriers.

rdar://104630103

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test source compatibility

atrick commented on lib/SILOptimizer/Utils/DeinitBarrierFinding.cpp:
```diff
@@ -0,0 +1,153 @@
+  Reachability::Result result;
+  Reachability reachability;
+
+  enum class Classification { Barrier, Other };
```
Why do we need `Classification` and `classifyInstruction` as opposed to just an `isBarrier`. It's a lot of code to understand just for one concept.

nate-chandler replied:
Thanks for pointing that out!  This code was adapted from a context where it made sense to have `Classification`.  It's not worth it here.

atrick commented on include/swift/SILOptimizer/Utils/DeinitBarrierFinding.h:
```diff
@@ -0,0 +1,53 @@
+
+} // namespace DeinitBarrierFinding
+
+void findDeinitBarriersAboveInstructions(
```
What makes this specific to deinit barriers? This looks like a simple adapter that converts Reachability's Visitor pattern to a simple `isBarrier` callback. That's pretty handy to avoid implementing a visitor. But can this utility just do that: take your `isBarrier`, wrap up a Visitor, and give you back a set of barriers?

The user needs to pass in a barrier function anyway, so can't they just pass in a function that takes a pointer to the Context and checks for deinit barriers?

nate-chandler committed:
[Reachability] Added findBarriersBackward.

nate-chandler committed:
[Test] Unit test lexical destroy folding.

Adapt a preexisting test to be a unit test that runs the utility.

nate-chandler committed:
[CanonOSSALifetime] Run on lexical lifetimes.

Previously, the utility bailed out on lexical lifetimes because it
didn't respect deinit barriers.  Here, deinit barriers are found and
added to liveness if the value is lexical.  This enables copies to be
propagated without hoisting destroys over deinit barriers.

rdar://104630103

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[Reachability] Removed restrictive assert.

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
[CanonOSSALifetime] Fixed term boundary extension.

If multiple terminators which branch to the same merge block are added
to the boundary, depending on whether a destroy_value can be found
within the block either (a) every terminator must be added to the
boundary or (b) the destroy_value must be added to the boundary exactly
once.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please benchmark

nate-chandler commented:
@swift-ci please test macos platform

nate-chandler commented:
@swift-ci please test macos platform

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

atrick reviewed (approved):
Awesome!

A couple of suggestions...

nate-chandler committed:
[Reachability] NFC: Allow more initial blocks.

Allow clients to specify any number (an array) of blocks beyond which
dataflow won't propagate rather than 1 or 0 (a pointer).

nate-chandler commented on lib/SILOptimizer/Analysis/Reachability.cpp:
```diff
@@ -0,0 +1,117 @@
+  reachability.findBarriers(*this);
+}
+
+void run(Context &context, ReachableBarriers &barriers,
```
Removed the namespace, renamed `FindBarriersBeforeInstructions::Dataflow` to `FindBarriersDataflow`, and restructured `Context` into the dataflow.

nate-chandler commented on include/swift/SILOptimizer/Analysis/Reachability.h:
```diff
@@ -887,6 +887,35 @@ void IterativeBackwardReachability<Effects>::Result::setEffectForBlock(
+  ReachableBarriers &operator=(ReachableBarriers const &) = delete;
+};
+
+void findBarriersBeforeInstructions(
```
> Does the defBlock need to dominate, or is it an optional non-barrier stopping point? Presumably this could be extended to multiple defBlocks, and maybe we should point that out somewhere

Took this opportunity to extend `IterativeBackwardReachability` to handle a list of `initialBlocks`.  

Renamed the function to `findBarriersBackward`.  It now takes a list of `initialBlocks` too.  And added a bit of documentation to the function declaration.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test
